### ISSUE TITLE
False positive unnecessary_safety_comment in doc test

### ISSUE THREAD

**Issue by TomFryersMidsummer** [rust-lang/rust-clippy#12048]
### Summary

Sometimes doc comments contain  `// Safety: ` notes. Perhaps something like
```
/// // Safety: & is dangerous
/// x = y & 0b111;
```
should raise the alarm, but if there is an unsafe block in the doc comment, then probably not – see example below. For some reason, it seems to require something above the function for the warning to trigger. I've used a pointless constant.

### Lint Name

unnecessary_safety_comment 

### Reproducer

I tried this code:

```rust
pub const X: u8 = 0;

/// Returns a pointer to five.
///
/// # Examples
///
/// ```
/// use foo::point_to_five;
///
/// let five_pointer = point_to_five();
/// // Safety: this pointer always points to a valid five.
/// let five = unsafe { *five_pointer };
/// assert_eq!(five, 5);
/// ```
pub fn point_to_five() -> *const u8 {
    static FIVE: u8 = 5;
    &FIVE
}
```

I saw this happen:

```
$ clippy-driver -D clippy::unnecessary_safety_comment foo.rs --crate-type lib
error: function has unnecessary safety comment
  --> foo.rs:15:1
   |
15 | pub fn point_to_five() -> *const u8 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: consider removing the safety comment
  --> foo.rs:11:1
   |
11 | /// // Safety: this pointer always points to a valid five.
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_safety_comment
   = note: requested on the command line with `-D clippy::unnecessary-safety-comment`

error: aborting due to 1 previous error
```

I expected to see this happen:

[no error]

### Version

```text
rustc 1.77.0-nightly (fb5ed726f 2023-12-28)
binary: rustc
commit-hash: fb5ed726f72c6d16c788517c60ec00d4564b9348
commit-date: 2023-12-28
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```
